### ISSUE TITLE
Logger doesn't print or write anything apart from the start and when a game starts/ends

### ISSUE THREAD

**Issue by AttackingOrDefending** [ShailChoksi/lichess-bot#290]
At info level, the logger only prints or writes
```
2021-03-15 09:19:52,308:
    .   _/|
    .  // o\
    .  || ._)  lichess-bot 1.2.0
    .  //__\
    .  )___(   Play on Lichess with a bot

2021-03-15 09:19:53,744: Welcome AttackingOrDefending!
2021-03-15 09:19:53,762: You're now connected to https://lichess.org/ and awaiting challenges.
2021-03-15 09:21:40,488: --- Process Used. Total Queued: 0. Total Used: 1
2021-03-15 09:23:27,726: +++ Process Free. Total Queued: 0. Total Used: 0
```
and at debug level, only
```
2021-03-15 09:41:57,724:
    .   _/|
    .  // o\
    .  || ._)  lichess-bot 1.2.0
    .  //__\
    .  )___(   Play on Lichess with a bot

2021-03-15 09:41:57,780: Starting new HTTPS connection (1): lichess.org:443
2021-03-15 09:41:58,372: https://lichess.org:443 "GET /api/account HTTP/1.1" 200 None
2021-03-15 09:41:58,372: Welcome AttackingOrDefending!
2021-03-15 09:41:58,380: You're now connected to https://lichess.org/ and awaiting challenges.
2021-03-15 09:42:03,493: Something went wrong. Game is starting and we don't have a queued process
2021-03-15 09:42:03,501: --- Process Used. Total Queued: 0. Total Used: 1
2021-03-15 09:46:50,982: +++ Process Free. Total Queued: 0. Total Used: 0
```
I expected the logger to print/write the moves it gets from the book and the engine stats (score, depth etc.)

**Comment by TheYoBots** [comment]
That’s surprising, I just updated and tested the latest changes (https://github.com/ShailChoksi/lichess-bot/commit/132ac8b7da12b9a1d856b342cb2b9f1184332153) and I receive the logs for each move. I even see the move number for each move.

**Comment by AttackingOrDefending** [comment]
This has been happening for at least 6 months (maybe even from when I made my bot), so I always just replaced logger.info with print in print_stats, get_book_move and the other functions. logger.info only prints very few lines, so it is still working. I tried both python 3.9 and 3.7, but I think this was also happening when I used python 3.6.

**Comment by MarkZH** [comment]
What operating system are you using?

What command line are you using to start lichess-bot?

**Comment by MarkZH** [comment]
I can confirm this issue occurs on Windows 10. The logging stops once lichess-bot.py reaches [line 121](https://github.com/ShailChoksi/lichess-bot/blob/132ac8b7da12b9a1d856b342cb2b9f1184332153/lichess-bot.py#L121)
```py
pool.apply_async(play_game, [li, game_id, control_queue, engine_factory, user_profile, config, challenge_queue])
```
and nothing inside the async process gets logged.

 These stack exchange posts may be relevant to fixing this:

https://stackoverflow.com/questions/26167873/python-logging-with-multiprocessing-in-windows
https://stackoverflow.com/questions/34724643/python-logging-with-multiprocessing-root-logger-different-in-windows

**Comment by MarkZH** [comment]
More clues (I honestly don't know what to do about this and am hoping some other python programmers see this): I put some `print(logger)` statements before the `apply_async()` call above and at the beginning of the `play_game()` function and got this output:
```
start() logger:
<Logger __main__ (DEBUG)>
play_game() logger:
<Logger __mp_main__ (WARNING)>
```
So the logger in the `play_game()` process doesn't inherit the logging level (or any other configuration) from the parent process. This is a Windows-specific bug.

**Comment by AttackingOrDefending** [comment]
I tried using `logger.setLevel(logging.INFO)` at the beginning of play_game and print_move_number and the output of `print(logger)` was `<Logger __mp_main__ (INFO)>` but it still didn't print anything. After I changed `logger.info` to `logger.warn` in print_move_number I saw `move: 1`.

**Comment by imBigo** [comment]
I am also getting this issue. Started happening for me after I started using [#281](https://github.com/ShailChoksi/lichess-bot/pull/281).

**Comment by MarkZH** [comment]
I have a pull request that should solve this issue. I would appreciate it if commenters here could test it.

**Comment by AttackingOrDefending** [comment]
It works for me. The output is
```
2021-03-18 12:37:16,042:
    .   _/|
    .  // o\
    .  || ._)  lichess-bot 1.2.0
    .  //__\
    .  )___(   Play on Lichess with a bot

2021-03-18 12:37:17,007: Welcome AttackingOrDefending!
2021-03-18 12:37:17,009: You're now connected to https://lichess.org/ and awaiting challenges.
2021-03-18 12:37:34,439: --- Process Used. Total Queued: 0. Total Used: 1
2021-03-18 12:37:37,512: +++ https://lichess.org/ru89knbg/white Bullet vs AI level 8
2021-03-18 12:37:37,584:
2021-03-18 12:37:37,585: move: 1
2021-03-18 12:37:37,587: Got move g1f3 from book engines/AOD.bin
2021-03-18 12:37:41,795:
2021-03-18 12:37:41,797: move: 2
2021-03-18 12:37:41,813: Got move d2d4 from book engines/full_original.bin
2021-03-18 12:37:44,156:
2021-03-18 12:37:44,157: move: 3
2021-03-18 12:37:44,172: Got move c2c4 from book engines/full_original.bin
2021-03-18 12:37:46,046:
2021-03-18 12:37:46,047: move: 4
2021-03-18 12:37:46,051: Got move b1c3 from book engines/full_original.bin
2021-03-18 12:37:48,091:
2021-03-18 12:37:48,092: move: 5
2021-03-18 12:37:48,097: Got move e2e3 from book engines/full_original.bin
2021-03-18 12:37:50,184:
2021-03-18 12:37:50,185: move: 6
2021-03-18 12:37:50,187: Searching for wtime 63575 btime 62620
2021-03-18 12:38:08,626:     depth: 23
2021-03-18 12:38:08,628:     nps: 304718
2021-03-18 12:38:08,628:     nodes: 5418817
2021-03-18 12:38:08,629:     score: PovScore(Cp(+4), WHITE)
2021-03-18 12:38:11,642:
2021-03-18 12:38:11,643: move: 7
2021-03-18 12:38:11,654: Searching for wtime 45948 btime 62960
2021-03-18 12:38:17,716:     depth: 19
2021-03-18 12:38:17,718:     nps: 198638
2021-03-18 12:38:17,719:     nodes: 1189251
2021-03-18 12:38:17,722:     score: PovScore(Cp(+27), WHITE)
2021-03-18 12:38:19,896:
2021-03-18 12:38:19,898: move: 8
2021-03-18 12:38:19,918: Got move d2c3 from book engines/full_original.bin
2021-03-18 12:38:22,201:
2021-03-18 12:38:22,202: move: 9
2021-03-18 12:38:22,215: Got move b2b3 from book engines/full_original.bin
2021-03-18 12:38:24,690:
2021-03-18 12:38:24,691: move: 10
2021-03-18 12:38:24,708: Searching for wtime 43762 btime 64590
2021-03-18 12:38:29,201:     depth: 18
2021-03-18 12:38:29,203:     nps: 149004
2021-03-18 12:38:29,205:     nodes: 641317
2021-03-18 12:38:29,296:     score: PovScore(Cp(+23), WHITE)
2021-03-18 12:38:32,120:
2021-03-18 12:38:32,121: move: 11
2021-03-18 12:38:32,137: Searching for wtime 40394 btime 64650
2021-03-18 12:38:38,017:     depth: 21
2021-03-18 12:38:38,018:     nps: 230287
2021-03-18 12:38:38,018:     nodes: 1342804
2021-03-18 12:38:38,019:     score: PovScore(Cp(+37), WHITE)
2021-03-18 12:38:40,406: --- https://lichess.org/ru89knbg/white Game over
2021-03-18 12:38:40,988: +++ Process Free. Total Queued: 0. Total Used: 0
```
The only differences with print are that logger.info also has the date and time and that it is now green instead of white.

**Comment by imBigo** [comment]
> I have a pull request that should solve this issue. I would appreciate it if commenters here could test it.

Fixed it for me.

**Comment by imBigo** [comment]
@MarkZH I have noticed a weird behavior. Sometimes logger will print 2x the same thing after playing a game. Here's an example:
```2021-03-18 10:54:51,636: move: 52
2021-03-18 10:54:51,636: move: 52
2021-03-18 10:54:51,636: Searching for wtime 1640 btime 12310
2021-03-18 10:54:51,636: Searching for wtime 1640 btime 12310
2021-03-18 10:54:51,655:     depth: 3
2021-03-18 10:54:51,655:     depth: 3
2021-03-18 10:54:51,655:     seldepth: 12
2021-03-18 10:54:51,655:     seldepth: 12
2021-03-18 10:54:51,656:     nps: 342666
2021-03-18 10:54:51,656:     nps: 342666
2021-03-18 10:54:51,656:     nodes: 2056
2021-03-18 10:54:51,656:     nodes: 2056
2021-03-18 10:54:51,656:     score: PovScore(Cp(0), WHITE)
2021-03-18 10:54:51,656:     score: PovScore(Cp(0), WHITE)
2021-03-18 10:54:51,657:     tbhits: 0
2021-03-18 10:54:51,657:     tbhits: 0
2021-03-18 10:54:51,657:     time: 0.015
2021-03-18 10:54:51,657:     time: 0.015
```

**Comment by MarkZH** [comment]
@imBigo When do the repetitions happen? Is it only for the last move of a game? Does this happen every game? Also, what operating system are you using?

**Comment by imBigo** [comment]
I'm using Windows 10. The repetition started after my bot won a game by timeout. After that repetition will not stop unless I restart lichess-bot.py

**Comment by MarkZH** [comment]
@imBigo Does the repetition happen often? I've just tried several consecutive games where my bot won variously on time and checkmate and I didn't see the repeating lines. I'm on Windows 10 as well.

**Comment by imBigo** [comment]
> @imBigo Does the repetition happen often? I've just tried several consecutive games where my bot won variously on time and checkmate and I didn't see the repeating lines. I'm on Windows 10 as well.

Yes it does. Left my bot online for a night and it was actually repeating each line almost 10 times. It happens to me each couple of games and just keeps on adding to the repetition amount.

**Comment by MarkZH** [comment]
@imBigo Since I can't seem to induce this behavior on my end, I'll have to debug by proxy. Please, replace the logging_listener_proc function (line 75) with this and post what gets printed when the repetitions start.
```py
def logging_listener_proc(queue, configurer, level, log_filename):
    configurer(level, log_filename)
    logger = logging.getLogger()
    while not terminated:
        try:
            print('Waiting for log entry')
            logger.handle(queue.get())
            print('Got log entry\n')
        except Exception:
            pass
```
The added `print()` statements should reveal whether this function is printing multiple times or if the `play_game()` logger is sending duplicate log entries.

**Comment by imBigo** [comment]
@MarkZH Here is what I get:
```
Waiting for log entry
2021-03-20 17:29:10,937: Searching for wtime 37060 btime 18189
Got log entry

Waiting for log entry
2021-03-20 17:29:10,937: Searching for wtime 37060 btime 18189
Got log entry
```

**Comment by MarkZH** [comment]
OK. So the `start_game()` logger is sending duplicate entries. Let's add some more `print()` statements to `game_logging_configurer()`:

```py
def game_logging_configurer(queue, level):
    if sys.platform == 'win32':
        h = logging.handlers.QueueHandler(queue)
        root = logging.getLogger()
        root.addHandler(h)
        root.setLevel(level)
        print(root)
        print(root.handlers)
```
It would be good to see what gets printed when logs are only printed once and when they are printed multiple times. If it doesn't take too long, it would be nice to see how this changes as more and more duplicate log lines get printed.

**Comment by imBigo** [comment]
@MarkZH Here is what I get with 2x repetition
```
2021-03-20 21:10:29,809: +++ Process Free. Total Queued: 0. Total Used: 0
2021-03-20 21:10:31,126:     Accept Blitz rated challenge from BOT LCzeta(2363)
2021-03-20 21:10:31,497: --- Process Queue. Total Queued: 1. Total Used: 0
2021-03-20 21:10:31,611: --- Process Used. Total Queued: 0. Total Used: 1
<RootLogger root (INFO)>
[<QueueHandler (NOTSET)>, <QueueHandler (NOTSET)>]
2021-03-20 21:10:32,287: +++ https://lichess.org/rAKVFCES/black Blitz vs BOT LCzeta(2363)
Got log entry

Waiting for log entry
2021-03-20 21:10:32,287: +++ https://lichess.org/rAKVFCES/black Blitz vs BOT LCzeta(2363)
Got log entry

Waiting for log entry
2021-03-20 21:10:39,781:
Got log entry

Waiting for log entry
2021-03-20 21:10:39,781:
Got log entry

Waiting for log entry
2021-03-20 21:10:39,781: move: 1
Got log entry

Waiting for log entry
2021-03-20 21:10:39,781: move: 1
Got log entry
```

**Comment by MarkZH** [comment]
Here's the problem: `[<QueueHandler (NOTSET)>, <QueueHandler (NOTSET)>]`

Duplicate handlers are being added to the logger for `play_game()` messages. Since each handler is called by `logger.handle(queue.get())` in `logging_listener_proc()`, the duplicate handlers perform duplicate actions.

Try adding the line `root.handlers.clear()` after `root = logging.getLogger()` and let me know if the duplication persists.

The function should now look like this:
```py
def game_logging_configurer(queue, level):
    if sys.platform == 'win32':
        h = logging.handlers.QueueHandler(queue)
        root = logging.getLogger()
        root.handlers.clear()
        root.addHandler(h)
        root.setLevel(level)
        print(root)
        print(root.handlers)
```

Out of curiousity, what distribution and version of Python are you using? I have standard Python 3.9, and when I tried deliberately calling `root.addHandler(h)` twice on my computer duplicates were not added.

**Comment by imBigo** [comment]
I'm using standard Python 3.9.2. Issue seems to be fixed with `root.handlers.clear()`

**Comment by MarkZH** [comment]
I've pushed a commit that adds the `root.handlers.clear()` line. That should be the final fix for this issue.

@imBigo One more question for my own curiosity: what command line shell do you use to start lichess-bot? Cmd, powershell, git bash, something else?

**Comment by imBigo** [comment]
> I've pushed a commit that adds the `root.handlers.clear()` line. That should be the final fix for this issue.
> 
> @imBigo Thanks for your work. One more question for my own curiosity: what command line shell do you use to start lichess-bot? Cmd, powershell, git bash, something else?

I use cmd: python3 lichess-bot.py

The problem might have been related to my system. I had some issues with my older python version after updating lichess-bot. I did all the normal steps and installed new pip versions etc. Bot would start but not play. Finally I decided to try `python3` and installed it (from the Microsoft store for some reason) and installed pip reqs, added to path etc. I'm thinking maybe it has something to do with that.

Thank **you** for your hard work, I didn't do much here :P
